src/node_quic_crypto.cc

```diff
@@ -713,56 +727,128 @@ Local<Value> GetServerName(
 }
 
 // Get the ALPN protocol identifier that was negotiated for the session
-Local<Value> GetALPNProtocol(Environment* env, SSL* ssl) {
+Local<Value> GetALPNProtocol(QuicSession* session) {
   Local<Value> alpn;
   const unsigned char* alpn_buf = nullptr;
   unsigned int alpnlen;
+  auto ctx = session->CryptoContext();
 
-  SSL_get0_alpn_selected(ssl, &alpn_buf, &alpnlen);
+  SSL_get0_alpn_selected(**ctx, &alpn_buf, &alpnlen);
   if (alpnlen == sizeof(NGTCP2_ALPN_H3) - 2 &&
       memcmp(alpn_buf, NGTCP2_ALPN_H3 + 1, sizeof(NGTCP2_ALPN_H3) - 2) == 0) {
-    alpn = env->quic_alpn_string();
+    alpn = session->env()->quic_alpn_string();
   } else {
-    alpn = OneByteString(env->isolate(), alpn_buf, alpnlen);
+    alpn = OneByteString(session->env()->isolate(), alpn_buf, alpnlen);
   }
   return alpn;
 }
 
-Local<Value> GetCipherName(Environment* env, SSL* ssl) {
+Local<Value> GetCertificate(QuicSession* session) {
+  crypto::ClearErrorOnReturn clear_error_on_return;
+  auto ctx = session->CryptoContext();
+  Local<Value> value = v8::Undefined(session->env()->isolate());
+  X509* cert = SSL_get_certificate(**ctx);
+  if (cert != nullptr)
+    value = crypto::X509ToObject(session->env(), cert);
+  return value;
+}
+
+Local<Value> GetEphemeralKey(QuicSession* session) {
+  Environment* env = session->env();
+  Local<Context> context = env->context();
+
+  Local<Object> info = Object::New(env->isolate());
+  auto ctx = session->CryptoContext();
+
+  EVP_PKEY* raw_key;
+  if (SSL_get_server_tmp_key(**ctx, &raw_key)) {
+    crypto::EVPKeyPointer key(raw_key);
+    int kid = EVP_PKEY_id(key.get());
+    switch (kid) {
+      case EVP_PKEY_DH:
+        info->Set(context, env->type_string(),
+                  FIXED_ONE_BYTE_STRING(env->isolate(), "DH")).FromJust();
+        info->Set(context, env->size_string(),
+                  Integer::New(env->isolate(), EVP_PKEY_bits(key.get())))
+            .FromJust();
+        break;
+      case EVP_PKEY_EC:
+      case EVP_PKEY_X25519:
+      case EVP_PKEY_X448:
+        {
+          const char* curve_name;
+          if (kid == EVP_PKEY_EC) {
+            EC_KEY* ec = EVP_PKEY_get1_EC_KEY(key.get());
+            int nid = EC_GROUP_get_curve_name(EC_KEY_get0_group(ec));
+            curve_name = OBJ_nid2sn(nid);
+            EC_KEY_free(ec);
+          } else {
+            curve_name = OBJ_nid2sn(kid);
+          }
+          USE(info->Set(context, env->type_string(),
+                    FIXED_ONE_BYTE_STRING(env->isolate(), "ECDH")));
+          USE(info->Set(context, env->name_string(),
+                    OneByteString(env->isolate(), curve_name)));
+          USE(info->Set(context, env->size_string(),
+                    Integer::New(env->isolate(), EVP_PKEY_bits(key.get()))));
```

Comment:
So … this mixes repeated use of `USE()` here (which isn’t great because it might mean calling into JS while an exception is pending) with `.FromJust()` above (which also isn’t great because we can do better than crash on exceptions).

I’d either pick `.FromJust()` and stick to that, or do proper error handling and go with
`if (info->Set(...).IsNothing()) return MaybeLocal<Value>();`.

Reply:
I'll go with FromJust for now but we can revisit this later